include/command_graph_generator.h

```diff
@@ -106,14 +107,91 @@ class command_graph_generator {
+	bool is_recording() const { return m_recorder != nullptr; }
+
+	/// Maps command DAG types to their record type.
+	template <typename Command>
+	using record_type_for_t = utils::type_switch_t<Command, push_command(push_command_record), await_push_command(await_push_command_record),
+	    reduction_command(reduction_command_record), epoch_command(epoch_command_record), horizon_command(horizon_command_record),
+	    execution_command(execution_command_record), fence_command(fence_command_record)>;
+
+	template <typename Command, typename... CtorParamsAndRecordWithFn, size_t... CtorParamIndices, size_t RecordWithFnIndex>
+	Command* create_command_internal(std::tuple<CtorParamsAndRecordWithFn...>&& ctor_params_and_record_with,
+	    std::index_sequence<CtorParamIndices...> /* ctor_param_indices */, std::index_sequence<RecordWithFnIndex> /* record_with_fn_index */) {
+		auto* const cmd = m_cdag.create<Command>(std::move(std::get<CtorParamIndices>(ctor_params_and_record_with))...);
 		m_current_cmd_batch.insert(cmd);
+
+		if(is_recording()) {
+			const auto& record_with = std::get<RecordWithFnIndex>(ctor_params_and_record_with);
+			[[maybe_unused]] bool recorded = false;
+			record_with([&](auto&&... debug_info) {
+				m_recorder->record_command(
+				    std::make_unique<record_type_for_t<Command>>(std::as_const(*cmd), std::forward<decltype(debug_info)>(debug_info)...));
+				recorded = true;
+			});
+			assert(recorded && "record_debug_info() not called within recording function");
+		}
 		return cmd;
 	}
 
+	/// Wrapper around command_graph::create that adds commands to the current batch.
+	/// Records the command if a recorder is present.
+	///
+	/// Invoke as
+	/// ```
+	/// create<command-type>(command-ctor-params,
+	///     [&](const auto record_debug_info) { return record_debug_info(command-record-additional-ctor-params)})
+	/// ```
+	template <typename Command, typename... CtorParamsAndRecordWithFn>
+	Command* create_command(CtorParamsAndRecordWithFn&&... args) {
+		constexpr auto n_args = sizeof...(CtorParamsAndRecordWithFn);
+		static_assert(n_args > 0);
+		return create_command_internal<Command>(
+		    std::forward_as_tuple(std::forward<CtorParamsAndRecordWithFn>(args)...), std::make_index_sequence<n_args - 1>{}, std::index_sequence<n_args - 1>{});
+	}
+
+	/// Adds a new dependency between two commands and records it if recording is enabled.
+	void add_dependency(abstract_command* const from, abstract_command* const to, dependency_kind kind, dependency_origin origin) {
+		m_cdag.add_dependency(from, to, kind, origin);
+		if(is_recording()) { m_recorder->record_dependency(command_dependency_record(to->get_cid(), from->get_cid(), kind, origin)); }
+	}
+
+	struct assigned_chunk {
+		node_id executed_on;
+		chunk<3> chnk;
+	};
+
+	using buffer_requirements_map = std::unordered_map<buffer_id, std::unordered_map<access_mode, region<3>>>;
```

Comment:
Great idea, this led to a bunch more simplifications! I ended up making it a vector of `struct { buffer_id, region, region }`, because we never need to look up requirements by buffer id (well, almost - except when adding reduction requirements, which is kind of a hack anyway).

Reply:
Nice!